Add tests for basic `private protected` functionality.

gafter opened the pull request:
PR for a first pass of tests added for testing `private protected`

gafter commented:
retest windows_debug_vs-integration_prtest please

jcouv commented on src/Compilers/CSharp/Test/Emit/PrivateProtected.cs:
```diff
@@ -0,0 +1,456 @@
+        private static readonly DesktopStrongNameProvider s_defaultProvider = new SigningTestHelpers.VirtualizedStrongNameProvider(ImmutableArray.Create<string>());
+
+        [Fact]
+        public void RejectIncimpatibleModifiers()
```
typo

gafter commented:
Follow-up tests under development will appear in https://github.com/gafter/roslyn/pull/22 until a PR on this repository is created, after this PR is merged.

gafter commented:
test windows_release_unit64_prtest please

gafter commented:
The failure was

## Regression

**Microsoft.CodeAnalysis.Editor.UnitTests.CodeGeneration.CodeGenerationTests+CSharp.RemoveAttributeWithMultipleAttributes** (from Roslyn.Services.Editor.UnitTests)

Failing for the past 1 build (Since Failed#26 )
Took 0.84 sec.
### Stacktrace

MESSAGE:
System.AggregateException : 3 exception(s) were thrown during test.\r\nNote: exceptions may have been thrown by another test running concurrently with\r\nthis test. This can happen with any tests that share the same ExportProvider.\r\nExamining individual exception stacks may help reveal the original test and source \r\nof the problem.\r\n\r\nException 0:\r\nSystem.NullReferenceException: Object reference not set to an instance of an object.\r\n at Microsoft.VisualStudio.Text.AdornmentLibrary.TextMarkers.Implementation.TextMarkerVisualManager.OnBatchedTagsChanged(Object sender, BatchedTagsChangedEventArgs e)\r\n at Microsoft.VisualStudio.Text.Utilities.GuardedOperations.RaiseEvent[TArgs](Object sender, EventHandler`1 eventHandlers, TArgs args)\r\n\r\nException 1:\r\nSystem.ArgumentException: The specified NormalizedSnapshotSpanCollection extends beyond the end of the TextSnapshot.\r\n at Microsoft.VisualStudio.Text.NormalizedSnapshotSpanCollection..ctor(ITextSnapshot snapshot, IList`1 spans)\r\n at Microsoft.VisualStudio.Text.Projection.Implementation.BufferGraph.MapDownToFirstMatch(SnapshotSpan span, SpanTrackingMode trackingMode, Predicate`1 match)\r\n at Microsoft.VisualStudio.Text.Projection.Implementation.BufferGraph.MapDownToBuffer(SnapshotSpan span, SpanTrackingMode trackingMode, ITextBuffer targetBuffer)\r\n at Microsoft.VisualStudio.Text.Implementation.MappingSpan.GetSpans(ITextBuffer targetBuffer)\r\n at Microsoft.VisualStudio.Text.Implementation.MappingSpan.GetSpans(ITextSnapshot targetSnapshot)\r\n at Microsoft.VisualStudio.Text.InterTextAdornmentSupport.Implementation.InterLineAdornmentManager.OnBatchedTagsChanged(Object sender, BatchedTagsChangedEventArgs e)\r\n at Microsoft.VisualStudio.Text.Utilities.GuardedOperations.RaiseEvent[TArgs](Object sender, EventHandler`1 eventHandlers, TArgs args)\r\n\r\nException 2:\r\nSystem.NullReferenceException: Object reference not set to an instance of an object.\r\n at Microsoft.VisualStudio.Text.InterTextAdornmentSupport.Implementation.InterLineAdornmentManager.PerformLayout(SnapshotPoint trackingPoint)\r\n at Microsoft.VisualStudio.Text.InterTextAdornmentSupport.Implementation.InterLineAdornmentManager.OnBatchedTagsChanged(Object sender, BatchedTagsChangedEventArgs e)\r\n at Microsoft.VisualStudio.Text.Utilities.GuardedOperations.RaiseEvent[TArgs](Object sender, EventHandler`1 eventHandlers, TArgs args)\r\n\r\n\r\n---- System.NullReferenceException : Object reference not set to an instance of an object.\r\n---- System.ArgumentException : The specified NormalizedSnapshotSpanCollection extends beyond the end of the TextSnapshot.\r\n---- System.NullReferenceException : Object reference not set to an instance of an object.
+++++++++++++++++++
STACK TRACE:
at Microsoft.CodeAnalysis.Editor.UnitTests.Workspaces.TestWorkspace.Dispose(Boolean finalize) at Microsoft.CodeAnalysis.Editor.UnitTests.CodeGeneration.CodeGenerationTests.TestContext.Dispose() in q:\roslyn\src\EditorFeatures\Test\CodeGeneration\CodeGenerationTests.cs:line 946 at Microsoft.CodeAnalysis.Editor.UnitTests.CodeGeneration.CodeGenerationTests.<TestRemoveAttributeAsync>d__14`1.MoveNext() in q:\roslyn\src\EditorFeatures\Test\CodeGeneration\CodeGenerationTests.cs:line 519 --- End of stack trace from previous location where exception was thrown --- at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess(Task task) at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task) at Microsoft.CodeAnalysis.Editor.UnitTests.CodeGeneration.CodeGenerationTests.CSharp.<RemoveAttributeWithMultipleAttributes>d__83.MoveNext() in q:\roslyn\src\EditorFeatures\Test\CodeGeneration\CodeGenerationTests.CSharp.cs:line 924 --- End of stack trace from previous location where exception was thrown --- at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess(Task task) at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task) --- End of stack trace from previous location where exception was thrown --- at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess(Task task) at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task) --- End of stack trace from previous location where exception was thrown --- at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess(Task task) at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task) ----- Inner Stack Trace #1 (System.NullReferenceException) ----- at Microsoft.VisualStudio.Text.AdornmentLibrary.TextMarkers.Implementation.TextMarkerVisualManager.OnBatchedTagsChanged(Object sender, BatchedTagsChangedEventArgs e) at Microsoft.VisualStudio.Text.Utilities.GuardedOperations.RaiseEvent[TArgs](Object sender, EventHandler`1 eventHandlers, TArgs args) ----- Inner Stack Trace #2 (System.ArgumentException) ----- at Microsoft.VisualStudio.Text.NormalizedSnapshotSpanCollection..ctor(ITextSnapshot snapshot, IList`1 spans) at Microsoft.VisualStudio.Text.Projection.Implementation.BufferGraph.MapDownToFirstMatch(SnapshotSpan span, SpanTrackingMode trackingMode, Predicate`1 match) at Microsoft.VisualStudio.Text.Projection.Implementation.BufferGraph.MapDownToBuffer(SnapshotSpan span, SpanTrackingMode trackingMode, ITextBuffer targetBuffer) at Microsoft.VisualStudio.Text.Implementation.MappingSpan.GetSpans(ITextBuffer targetBuffer) at Microsoft.VisualStudio.Text.Implementation.MappingSpan.GetSpans(ITextSnapshot targetSnapshot) at Microsoft.VisualStudio.Text.InterTextAdornmentSupport.Implementation.InterLineAdornmentManager.OnBatchedTagsChanged(Object sender, BatchedTagsChangedEventArgs e) at Microsoft.VisualStudio.Text.Utilities.GuardedOperations.RaiseEvent[TArgs](Object sender, EventHandler`1 eventHandlers, TArgs args) ----- Inner Stack Trace #3 (System.NullReferenceException) ----- at Microsoft.VisualStudio.Text.InterTextAdornmentSupport.Implementation.InterLineAdornmentManager.PerformLayout(SnapshotPoint trackingPoint) at Microsoft.VisualStudio.Text.InterTextAdornmentSupport.Implementation.InterLineAdornmentManager.OnBatchedTagsChanged(Object sender, BatchedTagsChangedEventArgs e) at Microsoft.VisualStudio.Text.Utilities.GuardedOperations.RaiseEvent[TArgs](Object sender, EventHandler`1 eventHandlers, TArgs args)

cston commented on src/Compilers/CSharp/Test/Emit/PrivateProtected.cs:
```diff
@@ -0,0 +1,456 @@
+@"class Class
+{
+    public int Prop1 { get; private protected set; }
+    protected internal int Prop2 { get; private protected set; }
```
`protected internal` is allowed?

cston commented on src/Compilers/CSharp/Test/Emit/PrivateProtected.cs:
```diff
@@ -0,0 +1,456 @@
+    private internal int Field1;
+    internal private int Field2;
+    private internal protected int Field3;
+    internal protected private int Field4;
```
Include `public`.

cston replied:
Test modifiers separated by other valid modifiers:
```
private static protected int Field1;
protected partial private int Field2;
```

cston commented on src/Compilers/CSharp/Test/Emit/PrivateProtected.cs:
```diff
@@ -0,0 +1,456 @@
+                //     static private protected int Field1 = 2;
+                Diagnostic(ErrorCode.ERR_ProtectedInStatic, "Field1").WithArguments("C.Field1").WithLocation(3, 34)
+                );
+        }
```
Test `private protected` member in `sealed` class. Should generate `CS0628`?

cston replied:
Test hiding an `abstract` method.
```
abstract class A
{
    internal abstract void F();
}
abstract class B : A
{
    private protected new void F() { } // CS0533
}
```

cston replied:
Test unimplemented inaccessible method.
A.dll:
```
public abstract class A
{
    private protected abstract void F();
}
```
B.dll (no IVT access):
```
class B : A // CS0534
{
}
```

cston commented on src/Compilers/CSharp/Test/Emit/PrivateProtected.cs:
```diff
@@ -0,0 +1,456 @@
+namespace Microsoft.CodeAnalysis.CSharp.UnitTests
+{
+    [CompilerTrait(CompilerFeature.PrivateProtected)]
+    public class PrivateProtected : CSharpTestBase
```
Are there tests for `ISymbol.DeclaredAccessibility`?

cston commented on src/Compilers/CSharp/Test/Emit/PrivateProtected.cs:
```diff
@@ -0,0 +1,456 @@
+@"class Class
+{
+    private protected int Prop1 { get; private protected set; }
+    private int Prop2 { get; private protected set; }
```
```
private protected int Prop3 { get; protected internal set; }
private protected int Prop4 { get; protected set; }
private protected int Prop5 { get; internal set; }
private protected int Prop6 { get; public set; }
```

cston commented on src/Compilers/VisualBasic/Portable/Binding/Binder_Utils.vb:
```diff
@@ -198,6 +199,9 @@ Namespace Microsoft.CodeAnalysis.VisualBasic
                     '  Note: Need to be present both MustInherit & NotInheritable for properly reporting #30926
                     foundModifiers = foundModifiers Or currentModifier
                 Else
+                    If currentModifier = SourceMemberFlags.Private Or currentModifier = SourceMemberFlags.Protected Then
```
`OrElse`

cston commented on src/Compilers/VisualBasic/Portable/Binding/Binder_Utils.vb:
```diff
@@ -210,6 +214,16 @@ Namespace Microsoft.CodeAnalysis.VisualBasic
+                                     currentVersion.GetErrorName(),
+                                     ErrorFactory.ErrorInfo(Microsoft.CodeAnalysis.VisualBasic.Syntax.InternalSyntax.FeatureExtensions.GetResourceId(feature)),
+                                     New VisualBasicRequiredLanguageVersion(requiredVersion))
+                End If
```
```
InternalSyntax.Parser.CheckFeatureAvailability(
    diagBag,
    privateProtectedToken.GetLocation(),
    DirectCast(privateProtectedToken.SyntaxTree, VisualBasicSyntaxTree).Options.LanguageVersion,
    InternalSyntax.Feature.PrivateProtected)
```

cston commented on src/Compilers/VisualBasic/Test/Emit/PrivateProtected.vb:
```diff
@@ -0,0 +1,567 @@
+BC30176: Only one of 'Public', 'Private', 'Protected', 'Friend', or 'Protected Friend' can be specified.
+    Friend Protected Private Field4 As Integer
+                     ~~~~~~~
+</errors>)
```
Test `Private Protected Public` so it's clear we're not just testing existing `Private` + `Protected Friend` behavior.

cston commented on src/Compilers/CSharp/Test/Symbol/Symbols/SymbolErrorTests.cs:
```diff
@@ -3357,10 +3357,22 @@ internal partial struct S { }
+            //var comp = DiagnosticsUtils.VerifyErrorsAndGetCompilationWithMscorlib(text,
+            //    new ErrorDescription { Code = (int)ErrorCode.ERR_PartialModifierConflict, Line = 3, Column = 30 },
+            //    new ErrorDescription { Code = (int)ErrorCode.ERR_PartialModifierConflict, Line = 9, Column = 32 },
+            //    new ErrorDescription { Code = (int)ErrorCode.ERR_PartialModifierConflict, Line = 12, Column = 32 });
```
Remove commented out code.

gafter commented:
@cston I've addressed all comments about the existing tests. I've added suggestions for additional tests to https://github.com/dotnet/roslyn/issues/19722 .

Do you have any further comments?
